### ISSUE TITLE
Pupil Player: Time Stamps of world cam video are mixed up

### ISSUE THREAD

**Issue by emolyzr** [pupil-labs/pupil#1203]
Hi,
here's a complete issue report. Unfortunately, I am not a python coder, but should be easy to solve?! Roughly 10-15% of my recorded data show that issue. Realy appreciate any help to fix or work around that bug!

Setup: Pupil Mobile App publishes frames to pupil capture via WIFI to a laptop running pupil capture. Pupil Capture records the streamed data locally on the laptop.

Symptom: Pupil Player says recording lasted 14000 Minutes, although there were only 30 minutes of recording (see screenshot). In the time series section there is huge gap between the first few data points (here 2550 frames) and the last data points (here about 30000 frames).

![screenshot from 2018-05-30 12-07-45](https://user-images.githubusercontent.com/35568338/40726153-694c73ae-6425-11e8-83ac-60a793505c5a.png)

Relevant section from Player Log File: 2018-05-30 13:10:58,701 - player - [ERROR] video_capture.file_backend: Videofile pts are not evenly spaced, pts to index conversion may fail and be inconsitent.

Possible Cause: time sync seems to change the time stamps during the recording.

Possible Solution: open "world_timestamps.npy" --> find the time point where time sync mixed up the time stamps --> adapt the first few time stamps to be in the same time range as the last few time stamps (probably just calculate the difference at the point where is the gap, and add that difference to the first few time stamps)

Link to "world_timestamps.npy": https://we.tl/y4jwLESw5t

Complete Dataset upon request available.

Many thanks!

**Comment by emolyzr** [comment]
Tried to do that using MATLAB (https://we.tl/4eXWLfHQ2D) and a python toolbox, however this produces an error when loading the dataset with the corrected timestamps.npy file.

2018-05-31 12:41:48,723 - player - [ERROR] launchables.player: Process Player crashed with trace:
Traceback (most recent call last):
  File "launchables/player.py", line 215, in player
  File "shared_modules/video_capture/__init__.py", line 64, in init_playback_source
  File "shared_modules/video_capture/file_backend.py", line 148, in __init__
AssertionError: Timestamps need to be instances of python float, got <class 'numpy.ndarray'>

**Comment by papr** [comment]
Hey @emolyzr 
I wrote a small script that tries to fix the above timestamp issues as much as possible. Unfortunately, it is not possible to enforce time monotonicity for all of your recordings.

https://gist.github.com/papr/d0e45707684ba9fa72fcff5a3e548d40

Usage: `python3.6 fix_timestamps.py path_including_recordings` were `path_including_recordings` is the path to the folder including the recordings whose timestamp files are to be fixed.

The script iterates over all recording folders in `path_including_recordings` and checks if the `*_timestamps.npy` files in them need fixing. Fixed timestamps are saved next to the broken timestamp files with the name `*_timestamps-fixed.npy`. Therefore you will have to replace the broken timestamp files with the fixed ones by renaming the files manually. Make sure to backup the broken files accordingly.

Keep in mind that the script is not a long-term solution. We will try to ind a solution such that these kind of gaps do not appear anymore.
